core/src/main/java/org/apache/iceberg/PartitionStatsUtil.java

```diff
@@ -45,22 +49,53 @@ private PartitionStatsUtil() {}
    * @return the collection of {@link PartitionStats}
+   * @deprecated since 1.9.0, will be removed in 1.10.0; use {@link #computeStats(Table, Snapshot,
+   *     Snapshot)} instead.
    */
+  @Deprecated
   public static Collection<PartitionStats> computeStats(Table table, Snapshot snapshot) {
-    Preconditions.checkArgument(table != null, "table cannot be null");
-    Preconditions.checkArgument(Partitioning.isPartitioned(table), "table must be partitioned");
-    Preconditions.checkArgument(snapshot != null, "snapshot cannot be null");
+    return computeStats(table, null, snapshot).values();
+  }
 
-    StructType partitionType = Partitioning.partitionType(table);
-    List<ManifestFile> manifests = snapshot.allManifests(table.io());
-    Queue<PartitionMap<PartitionStats>> statsByManifest = Queues.newConcurrentLinkedQueue();
-    Tasks.foreach(manifests)
-        .stopOnFailure()
-        .throwFailureWhenFinished()
-        .executeWith(ThreadPools.getWorkerPool())
-        .run(manifest -> statsByManifest.add(collectStats(table, manifest, partitionType)));
+  /**
+   * Computes the partition stats incrementally after the given snapshot to current snapshot. If the
+   * given snapshot is null, computes the stats completely instead of incrementally.
+   *
+   * @param table the table for which partition stats to be computed.
+   * @param fromSnapshot the snapshot after which partition stats is computed (exclusive).
+   * @param currentSnapshot the snapshot till which partition stats is computed (inclusive).
+   * @return the {@link PartitionMap} of {@link PartitionStats}
+   */
+  public static PartitionMap<PartitionStats> computeStats(
+      Table table, Snapshot fromSnapshot, Snapshot currentSnapshot) {
+    Preconditions.checkArgument(table != null, "Table cannot be null");
+    Preconditions.checkArgument(Partitioning.isPartitioned(table), "Table must be partitioned");
+    Preconditions.checkArgument(currentSnapshot != null, "Current snapshot cannot be null");
 
-    return mergeStats(statsByManifest, table.specs());
+    Predicate<ManifestFile> manifestFilePredicate = file -> true;
+    if (fromSnapshot != null) {
+      Preconditions.checkArgument(currentSnapshot != fromSnapshot, "Both the snapshots are same");
+      Preconditions.checkArgument(
+          SnapshotUtil.isAncestorOf(table, currentSnapshot.snapshotId(), fromSnapshot.snapshotId()),
+          "Starting snapshot %s is not an ancestor of current snapshot %s",
+          fromSnapshot.snapshotId(),
+          currentSnapshot.snapshotId());
+      Set<Long> snapshotIdsRange =
+          Sets.newHashSet(
+              SnapshotUtil.ancestorIdsBetween(
+                  currentSnapshot.snapshotId(), fromSnapshot.snapshotId(), table::snapshot));
+      manifestFilePredicate =
+          manifestFile ->
+              snapshotIdsRange.contains(manifestFile.snapshotId())
+                  && !manifestFile.hasExistingFiles();
```

Comment:
Don't we want this as a default predicate?
```
manifestFile -> !manifestFile.hasExistingFiles()
```

Reply:
we could add it as default filter:
````
List<ManifestFile> manifests =
  currentSnapshot.allManifests(table.io()).stream()
      .filter(manifestFilePredicate)
      .filter(manifestFile -> !manifestFile.hasExistingFiles())
      .collect(Collectors.toList());
````

Reply:
good point. 

While computing incremental, I observed that it may become duplicate counts. So, I added. 
I do have some gaps, I need to understand fully when and all we mark manifest entry as existing.
Is there any scenario exist to consider "existing" entries or just "added" is enough? 
 
There is another check down below, that considers both added and existing (added long back). 
https://github.com/apache/iceberg/blob/d54d81ecc1968eb0ec16a1f266e1d0b29e53b2ef/core/src/main/java/org/apache/iceberg/PartitionStatsUtil.java#L103

I will update the code to just keep added entry and also add a testcase of rewrite data files to ensure stats are same after the rewrite.

Reply:
what if we have compaction and expire snapshots? new manifests would have the EXISTING entries?

Reply:
> I will update the code to just keep added entry and also add a testcase of rewrite data files to ensure stats are same after the rewrite.

I remember now why I don't have this testcase, compaction was only supported at spark layer and this testcase is in core/data module. So, I was waiting for spark procedure to add these test case.

Reply:
ok. I spent some time. 

The base code behavior of keeping existing entries is also proper. Because when the user calls merge append. It marks old manifest as delete and creates a new manifest which has existing entries + new entries. So, we need to consider both.

Reply:
I don't completely understand why it should be any different, since for incremental processing, we only cover the snapshot range (e.g., [s2...s5]). 
If we had performed a recompute, we would have included all the manifest files.

Reply:
I have removed the existing file level check from predicate and handled at entry level

```
if (entry.isLive()) {
          // Live can have both added and existing entries. Don't consider existing entries for
          // incremental compute as it was already included in previous compute.
          if (!incremental || entry.status() == ManifestEntry.Status.ADDED) {
            stats.liveEntry(file, snapshot);
          }
        } else {
          stats.deletedEntry(snapshot);
        }
```

This way existing test of both incremental and full compute will pass.

Reply:
compaction doesn't remove files. if we do snapshot expiry after compaction - stats is dropped as well so we start fresh.

Reply:
compaction doesn't remove the data. if we. expire S1 and S2 we don't have prev stats and start fresh (i.e. full compute)

Reply:
If we don't expire data, could we detect that S3 is only a compaction commit, and the stats don't need to be changed?

What if S3 instead is a MoW commit? Can we detect the changes and calculate stats incrementally?

Reply:
We didn't need decrement stats for full compute because we were discarding the deleted manifests. Only considering live manifests. 

Now, I am not really sure for compaction, the current code will work. We may need decrement stats just for incremental compute. I will test compaction scenario tomorrow and handle this.